tests/test_page.py

```diff
@@ -395,3 +396,142 @@ async def test_expose_function_should_work_with_complex_objects(page, server):
+        )
+
+
+@pytest.mark.skip_browser("firefox")
+async def test_page_error_should_handle_object(page, is_chromium):
+    # Firefox just does not report this error.
+    [error, _] = await asyncio.gather(
+        page.waitForEvent("pageerror"),
+        page.evaluate("() => setTimeout(() => { throw {}; }, 0)"),
+    )
+    assert error.message == "Object" if is_chromium else "[object Object]"
+
+
+@pytest.mark.skip_browser("firefox")
+async def test_page_error_should_handle_window(page, is_chromium):
+    # Firefox just does not report this error.
+    [error, _] = await asyncio.gather(
+        page.waitForEvent("pageerror"),
+        page.evaluate("() => setTimeout(() => { throw window; }, 0)"),
+    )
+    assert error.message == "Window" if is_chromium else "[object Window]"
+
+
+expected_output = "<html><head></head><body><div>hello</div></body></html>"
+
+
+async def test_set_content_should_work(page, server):
+    await page.setContent("<div>hello</div>")
+    result = await page.content()
+    assert result == expected_output
+
+
+async def test_set_content_should_work_with_domcontentloaded(page, server):
+    await page.setContent("<div>hello</div>", waitUntil="domcontentloaded")
+    result = await page.content()
+    assert result == expected_output
+
+
+async def test_set_content_should_work_with_doctype(page, server):
+    doctype = "<!DOCTYPE html>"
+    await page.setContent(f"{doctype}<div>hello</div>")
+    result = await page.content()
+    assert result == f"{doctype}{expected_output}"
+
+
+async def test_set_content_should_work_with_HTML_4_doctype(page, server):
+    doctype = '<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">'
+    await page.setContent(f"{doctype}<div>hello</div>")
+    result = await page.content()
+    assert result == f"{doctype}{expected_output}"
+
+
+async def test_set_content_should_respect_timeout(page, server):
+    img_path = "/img.png"
+    # stall for image
+    server.set_route(img_path, lambda request: None)
+    error = None
+    try:
+        await page.setContent(
+            '<img src="${server.PREFIX + img_path}"></img>', timeout=1
+        )
+    except Error as e:
+        error = e
+    assert isinstance(error, TimeoutError)
+
+
+async def test_set_content_should_respect_default_navigation_timeout(page, server):
+    page.setDefaultNavigationTimeout(1)
+    img_path = "/img.png"
+    # stall for image
+    await page.route(img_path, lambda route, request: None)
+
+    error = None
+    try:
+        await page.setContent(f'<img src="{server.PREFIX + img_path}"></img>')
+    except Error as e:
+        error = e
+    assert "Timeout 1ms exceeded during" in error.message
+    assert isinstance(error, TimeoutError)
+
+
+async def test_set_content_should_await_resources_to_load(page, server):
+    img_path = "/img.png"
+    img_route = page._scope._loop.create_future()
```

Comment:
Is there a difference to asyncio.Futufe()?

Reply:
Doc is saying

```This is the preferred way to create Futures in asyncio. This lets third-party event loops provide alternative implementations of the Future object (with better performance or instrumentation)```